src/Microsoft.AspNetCore.Sockets/Transports/WebSocketsTransport.cs

```diff
@@ -152,12 +152,13 @@ private void LogFrame(string action, WebSocketFrame frame)
 
         private async Task StartSending(IWebSocketConnection ws)
         {
-            while (true)
+            while (await _application.Input.WaitToReadAsync())
             {
                 // Get a frame from the application
-                try
+                Message message;
+                if (_application.Input.TryRead(out message))
```

Comment:
We should update this pattern to use a loop here. `WaitToReadAsync` returns `Task<T>` so we should avoid allocating a new Task until we actually have a high likelihood of blocking. We should try to synchronously gather all the messages we can until we hit the end.